Allow setting volume on Ring devices

daniel-k opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Turn volume sensor entities of the Ring integration into number entities so they become settable. This will require deactivating the old sensor entities so the new number entities can be created automatically.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
So far, volume from Ring devices was read-only. The underlying library already supports setting the volume, so this PR proposes to change the volume sensor entities into number entities so that we can change the volume from Home Assistant.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

This is my first contribution to HA and the Ring integration in particular, so please bear with me :upside_down_face:  While linters are passing, I didn't make any changes to the tests so far. Some guidance would be welcome.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
  - https://github.com/home-assistant/home-assistant.io/pull/34766 

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

daniel-k committed:
Turn Ring Doorbell and Chime volumes into number entities.

daniel-k committed:
turn RingOther volumes into numbers as well

daniel-k committed:
fix linter issues

daniel-k committed:
move other volume strings into `number` section

daniel-k commented:
@sdb9696 
I would like to add tests, but I don't understand where the fixture are coming from. For instance, where is `sensor.downstairs_volume` defined so that I can turn that into a number?

FYI: I have tested the changes locally with a real Chime and Doorbell, I don't have an Intercom (`RingOther`) unfortunately.

sdb9696 commented:
> @sdb9696 I would like to add tests, but I don't understand where the fixture are coming from. For instance, where is `sensor.downstairs_volume` defined so that I can turn that into a number?
> 
> FYI: I have tested the changes locally with a real Chime and Doorbell, I don't have an Intercom (`RingOther`) unfortunately.

Hi Daniel, thanks for the PR, it's looking pretty good.  The fixtures are in https://github.com/home-assistant/core/blob/dev/tests/components/ring/fixtures/devices.json. You'll need to do something similar to [this](https://github.com/home-assistant/core/pull/125775/files#diff-b487e8bddceac9e066a21d6090120f137d32187383b9a18ae4f2a0abe0f531f1R157-R163) to get the mock working with updates.

sdb9696 commented:
Also we can't just delete the sensors, they need to be deprecated. If you look these other sensors you should see how.

https://github.com/home-assistant/core/blob/75d3ea34fcd3365be1691f3449e11d9dc315f6d4/homeassistant/components/ring/sensor.py#L190

daniel-k committed:
Merge remote-tracking branch 'origin/dev' into feature/ring-set-volume

daniel-k committed:
add back old volume sensors but deprecate them

daniel-k committed:
add tests for `ring.number`

daniel-k committed:
add back strings for sensors that have just become deprecated

daniel-k committed:
remove deprecated volume sensors from test

daniel-k commented:
> Also we can't just delete the sensors, they need to be deprecated. If you look these other sensors you should see how.

Awesome, thanks for the pointer! See 68abeb7812a6de951686399a56ba2424ea327dff, bdb5a1ebe143afb871dc3cc889e13cda66a9c360 and fc95af66e7136202dca9560325d88b811ec22c45

I've added some tests in 8651208763ad60bff8d4c266497504ef4f762d0f. Hope that's enough, because other more in-depth tests for the other platform still look quite daunting to me :sweat_smile:

sdb9696 commented:
Hey @daniel-k  I’m on my mob so apols if I’ve missed something but you should keep the tests for the deprecated. Example in binary sensor of how to create them in the entity_registry beforehand should be helpful. You could also create a fixture for this so you don’t have to write the same code over and over. I have an example in another PR I have open to migrate siren if you want to look.

daniel-k committed:
Revert "remove deprecated volume sensors from test"

This reverts commit fc95af66e7136202dca9560325d88b811ec22c45.

daniel-k committed:
create entities for deprecated sensors so that tests still run

daniel-k commented:
No worries, here's my next attempt :) 1e01054dc5

ALthough I feel like there's a bit of redundancy in the tests across platforms regarding depecrated entities. Might be worth looking into refactoring that into some common tests.

sdb9696 commented:
So you’re right about the tests but I’ll come back on that tomorrow. I notice that the CI is failing. Are you running `pre-commit` checks? You should.

daniel-k committed:
remove print

daniel-k commented:
> I notice that the CI is failing. Are you running `pre-commit` checks? You should.

I ran them manually and obviously forgot to run the checks on my last commit, sorry. I've properly set up the pre-commit hook now and fixed the issue in 5a0b235.

sdb9696 reviewed (changes requested):
Hey, looking good, a few tweaks needed.

Regarding the duplication of logic around the tests for deprecated entities I agree.  I had originally just [patched the deprecated check method](https://github.com/home-assistant/core/pull/124879#discussion_r1748943588) but it's preferred not to patch the integration itself. I have an [architecture discussion](https://github.com/home-assistant/architecture/discussions/1133) open to centralise the deprecation logic and I will add a section to it regarding testing as it's an interesting question as to how to centralise it and deal with missing entities etc. Especially if snapshot's are being used.

sdb9696 commented:
The new platform will require a documentation PR linked in the "additional information" section (which you can get back from another PR).  See https://github.com/home-assistant/home-assistant.io/pull/34581 for an example of how the docs need to change for the new platform.

New platforms generally have to be isolated from other changes in a single PR. In this instance I think it is ok to include the deprecation as it's connected to the new platform and it's just one liners on the entity descriptions. That's an opinion though and a core member may disagree.

daniel-k committed:
add entities immediately

daniel-k committed:
move `RingNumberEntityDescription` above `RingNumber` and remove unused import

daniel-k committed:
remove irrelevant comment about history

daniel-k committed:
fix not using `setter_fn`

daniel-k committed:
add missing icons for other volume entities

daniel-k committed:
rename `entity` -> `entity_id` in number tests

daniel-k committed:
fix typing in number test

daniel-k committed:
use constants for `hass.services.async_call()`

daniel-k commented on tests/components/ring/test_number.py:
```diff
@@ -0,0 +1,94 @@
+        "number",
+        "set_value",
+        {"entity_id": entity, "value": new_value},
+        blocking=True,
```
See e2d2a9dfdd

daniel-k commented on tests/components/ring/test_number.py:
```diff
@@ -0,0 +1,94 @@
+    """Tests that the devices are registered in the entity registry."""
+    await setup_platform(hass, Platform.NUMBER)
+
+    entry = entity_registry.async_get(entity)
```
1df09e7b65

daniel-k commented on tests/components/ring/test_number.py:
```diff
@@ -0,0 +1,94 @@
+    entity_registry: er.EntityRegistry,
+    mock_ring_client,
+    entity,
+    unique_id,
```
3af64af837

daniel-k commented on tests/components/ring/test_number.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@pytest.mark.parametrize(
+    ("entity", "unique_id"),
```
1df09e7b65

daniel-k commented on homeassistant/components/ring/icons.json:
```diff
@@ -1,5 +1,10 @@
+      "volume": {
+        "default": "mdi:bell-ring"
+      }
+    },
```
5f6260f8ba

daniel-k commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+    async def async_set_native_value(self, value: float) -> None:
+        """Call setter on Ring device."""
+        async_setter = getattr(self._device, f"async_set_{self.entity_description.key}")
+        await async_setter(int(value))
```
Yup, good catch! I noticed the same when I saw your review yesterday :) Forgot to adapt this part after refactoring. Fixed in 23deff562e

daniel-k commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+
+    def _update_native_value(self) -> None:
+        # History values can drop off the last 10 events so only update
+        # the value if it's not None
```
096dccd9b1

daniel-k commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+        if description.exists_fn(device)
+    ]
+
+    async_add_entities(entities)
```
ad988f8218

daniel-k commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+"""Component providing HA number support for Ring Door Bell/Chimes."""
+
+from __future__ import annotations
```
Oh yeah, I didn't realize that. The whole structure was copy & paste from `switch.py` I think :thinking:  Fixed in 4da13d3e28

daniel-k committed:
Merge remote-tracking branch 'origin/dev' into feature/ring-set-volume

sdb9696 commented:
Hey @daniel-k, will go through these tomorrow. Just a quick note on the review process, it's up to you but I see you've been doing a commit per suggestion. You don't **have** to do that if it's too laborious. It's fine to do a single commit with all the tweaks and then you can mark the suggestions as resolved if it's as per the suggestion. If it's resolved differently or you disagree then leave it unresolved with a comment and I'll either reply again or agree with it and resolve it.

sdb9696 commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+        setter_fn=lambda device, value: device.async_set_volume(int(value)),
+        exists_fn=lambda device: isinstance(device, RingChime),
+    ),
+    RingNumberEntityDescription[RingDoorBell](
+        key="volume",
```
It would be great if you'd submit a PR to the library as I don't have a chime myself to test with (and it'll be tiny). It might also encourage you to add more amazing features to the lib 😄

daniel-k replied:
Haha, fair enough. Here you go: https://github.com/python-ring-doorbell/python-ring-doorbell/pull/439

daniel-k commented:
> it's up to you but I see you've been doing a commit per suggestion. You don't **have** to do that if it's too laborious. It's fine to do a single commit with all the tweaks and then you can mark the suggestions as resolved if it's as per the suggestion.

No worries, I'm used to doing anyways. The changes are so small and not dependent on each other. This makes your life easier when reviewing (I hope). At least I personally like this way of communicating back the changes that were made.



> will go through these tomorrow

No worries! I'm already more than happy that you reviewed it so quickly and even more, you gave super helpful pointers and suggestions! :) I'm still finishing up with the `refresh_after` thing and when I'm done, I'll re-request review.

daniel-k committed:
use `@refresh_after` decorator instead of delaying updates manually

daniel-k commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+        super().__init__(device, coordinator)
+        self.entity_description = description
+        self._attr_unique_id = f"{device.id}-{description.key}"
+        self._no_updates_until = time.monotonic()
```
Awesome, that's good timing that #125775 just got merged. Here are my changes: cf98410b18. I hope I didn't miss anything. FWIW, tests are passing and setting volume still seems to work.

sdb9696 replied:
Looks good to me

daniel-k commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+        """Call update method."""
+
+        if self._no_updates_until > time.monotonic():
+            return
```
cf98410b18

daniel-k commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+        async_setter = getattr(self._device, f"async_set_{self.entity_description.key}")
+        await async_setter(int(value))
+
+        self._no_updates_until = time.monotonic() + SKIP_UPDATES_DELAY_SECONDS
```
cf98410b18

sdb9696 commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,152 @@
+        self.async_write_ha_state()
+
+
+NUMBER_TYPES: tuple[RingNumberEntityDescription[Any], ...] = (
```
Apologies I should have also said to bring this up before the `RingNumber`. That way the declarative items are all at the top of the file.

daniel-k replied:
There it goes! dd3d319dc6

daniel-k commented on tests/components/ring/test_number.py:
```diff
@@ -0,0 +1,94 @@
+        ("number.ingress_voice_volume", "11.0"),
+    ],
+)
+async def test_initial_state(
```
Ugh, that appears to be a bit too much black magic for me :sweat_smile: Honestly, I don't understand how it works. I tried copying the `test_states()` function (obviously replacing `SIREN` by `NUMBER`), but the command you suggested fails with `FAILED tests/components/ring/test_number.py - Failed: In test_states: function uses no argument 'entity_id'`. Since I really don't know what I'm doing it's hard for me to work eith this error message.

Can you provide a another pointer maybe? Or maybe this could be cleaned up later?

sdb9696 replied:
It should work fine. Maybe you just need to recompile your translations:
```python3 -m script.translations develop --all```

daniel-k replied:
Ah well, I just forgot to uncomment the `@pytest.mark.parametrize` decorator of the old function :sweat_smile:  Now the test passes, but tbh I can't really tell if the test makes sense or not. Here's what I came up with: 258c02e007

sdb9696 replied:
You've forgotten to include the generated `test_number.ambr` file

daniel-k replied:
Oops. d8999541ab12fefca7631f415ac1f0a774829a11

daniel-k committed:
move descriptors above entity class

daniel-k committed:
Merge branch 'dev' into feature/ring-set-volume

sdb9696 commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,166 @@
+        # the value if it's not None
+        native_value = self.entity_description.value_fn(self._device)
+        if native_value is not None:
+            self._attr_native_value = float(native_value)
```
ok

daniel-k committed:
Use snapshot to test states.

daniel-k committed:
Merge remote-tracking branch 'origin/dev' into feature/ring-set-volume

daniel-k commented:
@sdb9696 I think I have addressed every comment of yours now and I have opened a documentation PR here: https://github.com/home-assistant/home-assistant.io/pull/34766

Therefore, I've removed the draft status that @joostlek had set. I hope that's alright :)

daniel-k committed:
add missing snapshot file for number platform

sdb9696 commented on homeassistant/components/ring/number.py:
```diff
@@ -0,0 +1,152 @@
+from .coordinator import RingDataCoordinator
+from .entity import RingDeviceT, RingEntity, refresh_after
+
+SKIP_UPDATES_DELAY_SECONDS = 5
+
```
```suggestion
```

sdb9696 replied:
Just did a final check and spotted a variable left behind. Other than that I think this is good to go so approving on the basis that you'll fix this before merge happens.

daniel-k replied:
Ah, good catch! I missed it when moving to the new `refresh_after`.

sdb9696 reviewed (approved):
LGTM, thanks for the PR @daniel-k!

daniel-k committed:
Update homeassistant/components/ring/number.py

Co-authored-by: Steven B. <51370195+sdb9696@users.noreply.github.com>
